src/main/java/svnserver/ext/gitea/auth/GiteaUserDB.java

```diff
@@ -0,0 +1,161 @@
+
+  @NotNull
+  private final Collection<Authenticator> authenticators = Collections.singleton(new PlainAuthenticator(this));
+  @NotNull
+  private final GiteaContext context;
+  
+  GiteaUserDB(@NotNull SharedContext context) {
+    this.context = context.sure(GiteaContext.class);
+  }
+
+  @NotNull
+  @Override
+  public Collection<Authenticator> authenticators() {
+    return authenticators;
+  }
+
+  @Override
+  public void updateEnvironment(@NotNull Map<String, String> env, @NotNull User userInfo) {
+    final String externalId = userInfo.getExternalId();
+    env.put("SSH_ORIGINAL_COMMAND", "git");
+    env.put("GITEA_PUSHER_EMAIL", userInfo.getEmail());
+    if (externalId != null) {
+      env.put("GL_ID", PREFIX_USER + externalId);
+      env.put("GITEA_PUSHER_ID", userInfo.getExternalId());
+    }
+  }
+
+  @Override
+  public User check(@NotNull String userName, @NotNull String password) {
+    try {
+      final ApiClient apiClient = context.connect(userName, password);
+      threadLocal.set(apiClient);
+      final UserApi userApi = new UserApi(apiClient);
+      return createUser(userApi.userGetCurrent());
+    } catch (ApiException e) {
+      if (e.getCode() == HttpURLConnection.HTTP_UNAUTHORIZED || e.getCode() == HttpURLConnection.HTTP_FORBIDDEN) {
+        return null;
+      }
+      log.warn("User password check error: " + userName, e);
+      return null;
+    } catch (IOException e) {
+      log.warn("User password check error: " + userName, e);
+      return null;
+    }
+  }
+
+  public static ApiClient getCurrentApiClient(@NotNull String userName) {
+    ApiClient apiClient = threadLocal.get();
+    if (apiClient != null) {
+      for (Authentication auth : apiClient.getAuthentications().values()) {
+        if (auth instanceof HttpBasicAuth) {
+            HttpBasicAuth httpBasicAuth = (HttpBasicAuth) auth;
+            if (userName.equals(httpBasicAuth.getUsername())) {
+              return apiClient;
+            }
+        }
+      }
+    }
+    return null;
+  }
+
+  @NotNull
+  private User createUser(@NotNull io.gitea.model.User user) {
+    return User.create(user.getLogin(), user.getFullName(), user.getEmail(), "" + user.getId());
+  }
+
+  @Nullable
+  @Override
+  public User lookupByUserName(@NotNull String userName) {
+    ApiClient apiClient = context.connect();
+    try {
+      UserApi userApi = new UserApi(apiClient);
+      io.gitea.model.User user = userApi.userGet(userName);
+      return createUser(user);
+    } catch (ApiException e) {
+      if (e.getCode() != 404) {
+        log.warn("User lookup by name error: " + userName, e);
+      }
+      return null;
+    }
+  }
+
+  @Nullable
+  @Override
+  public User lookupByExternal(@NotNull String external) {
+    final String userId = removePrefix(external, PREFIX_USER);
+    if (userId != null) {
+      try {
+        final UserApi userApi = new UserApi(context.connect());
+        UserSearchList users = userApi.userSearch(null, null);
```

Comment:
Does this download *all* users?

Reply:
Yes. Unfortunately, in the current gitea API there is no way to search for user by ID.

However, lookupByExternal is only called by LfsAuthHelper, LfsRpc - neither of which I have actually properly looked at yet.

Reply:
I've added a pull request for a way to avoid requesting all users https://github.com/go-gitea/gitea/pull/4876